pipewire webrtc screencasting

betaboon opened the pull request:
in order to get webrtc working with pipewire there are several things to be done.

some are done in this PR, some others have to be done be the user:

done in this PR:
- address webrtc-downstream-bug (has to be removed when [this](https://github.com/emersion/xdg-desktop-portal-wlr/issues/23) is done)
- use more recent versions of `xdg-desktop-portal` and `pipewire` (so fix a segfault i was running into)
- provide `pipewire_0_2` to chrome (done here with a wrapper)

to do by the user:
- set `XDG_CURRENT_DESKTOP=sway` environment variable
- enable a specific chrome-flag (`chrome://flags/#enable-webrtc-pipewire-capturer`)

up for discussion:
- is it feasible to provide a wrapper for chrome to provide the library? this would have to be done for chromium and vivaldi as well
- or should we rely on the user to set `LD_LIBRARY_PATH` globaly ?
- the chrome-flag could be provided with the wrapper as well (with `--enable-features=WebRTCPipeWireCapturer`)
- same applies `XDG_CURRENT_DESKTOP`

colemickens commented:
This looks good. Haven't tested yet but an initial ask -- can you move the temp packages (pipewire/xdg-desktop-portal) to `pkgs-temp`, and go ahead and send (or link to) PRs for `pipewire` and `xdg-desktop-portal` upstream? Thanks!

betaboon commented:
i will create PRs for `pipewire` and `xdg-deskop-portal` today or tomorrow and think them here when done.

still the question of to-wrap-or-not-to-wrap is standing.
i see three options now:
- wrap every application that needs `pipewire_0_2` seperately (pro: cleaner con: more tedious)
- have the user set LD_LIBRARY_PATH via `programs.sway.extraSessionCommands`
- have the user set LD_LIBRARY_PATH via `environment.variables`

danshick commented:
Our pixel format flag has been removed and we've released 0.1.0.

colemickens commented:
@danshick cool; thanks!

@betaboon Do we know the timeframe for Chromium (+ presumably Electron) stuff to move to pipewire 0.3? I don't like any of the options, really, I'm not sure which is better.

danshick commented:
It's all the same WebRTC codebase. Someone wrote a 0.3 patch but was waiting on Chromium to approve their contributor agreement.

betaboon commented:
@colemickens I can't find any information on when chromium might use pipewire 0.3

- i removed `pipewire` from this PR as it already landed in nixpkgs-unstable
- i removed the patch for the alpha-flag from this PR
- i created a PR for [`xdg-desktop-portal`](https://github.com/NixOS/nixpkgs/pull/88598) in nixpkgs

berbiche commented:
Hi, is there a way to test screencasting? I am really interested in making this work in my configuration.

colemickens commented:
@berbiche see all of the comments in this thread: https://github.com/NixOS/nixpkgs/pull/84233

Particularly the top and bottom ones. They describe how you can pull my build of 'firefox' and test with it.

colemickens commented:
this has been available in nixpkgs for sometime, so I'm going to close this. Thanks!

betaboon commented:
very nice. thank you.
